content/integrate/prometheus-with-redis-enterprise/observability.md

```diff
@@ -0,0 +1,636 @@
+
+
+| Total Cores | Redis (ROR) | Redis on Flash (ROF) |
+| ------ | ------ | :------ |
+|1|1|1 |
+|4|3|3 |
+|8|5|3 |
+|12|8|4 |
+|16|10|5 |
+|32|24|10 |
+|64/96|32|20 |
+|128|32|32 |
+
+
+## Data access anti-patterns
+
+There are three data access patterns that can limit the performance of your Redis database:
+
+* Slow operations
+* Hot keys
+* Large keys
+
+This section defines each of these patterns and describes how to diagnose and mitigate them.
+
+## Slow operations
+
+**Slow operations** are operations that take longer than a few milliseconds to complete.
+
+Not all Redis operations are equally efficient.
+The most efficient Redis operations are O(1) operations; that is, they have a constant time complexity.
+Example of such operations include [GET]({{< relref "/commands/get" >}}),
+[SET]({{< relref "/commands/set" >}}), [SADD]({{< relref "/commands/sadd" >}}),
+and [HSET]({{< relref "/commands/hset" >}}).
+
+These constant time operations are unlikely to cause high CPU utilization. **Note:** Even so,
+it's still possible for a high rate of constant time operations to overwhelm an underprovisioned database.
+
+Other Redis operations exhibit greater levels of time complexity.
+O(n) (linear time) operations are more likely to cause high CPU utilization.
+Examples include [HGETALL]({{< relref "/commands/hgetall" >}}), [SMEMBERS]({{< relref "/commands/smembers" >}}),
+and [LREM]({{< relref "/commands/lrem" >}}).
+These operations are not necessarily problematic, but they can be if executed against data structures holding
+a large number of elements (for example, a list with 1 million elements).
+
+However, the [KEYS]({{< relref "/commands/keys" >}}) command should almost never be run against a
+production system, since returning a list of all keys in a large Redis database can cause significant slowdowns
+and block other operations. If you need to scan the keyspace, especially in a production cluster, always use the
+[SCAN]({{< relref "/commands/scan" >}}) command instead.
+
+### Troubleshooting
+
+The best way to discover slow operations is to view the slow log.
+The slow log is available in the Redis Enterprise and Redis Cloud consoles:
+* [Redis Enterprise slow log docs](https://redis.io/docs/latest/operate/rs/clusters/logging/redis-slow-log/)
+* [Redis Cloud slow log docs](https://redis.io/docs/latest/operate/rc/databases/view-edit-database/#other-actions-and-info)
+
+Redis Cloud dashboard showing slow database operations
+{{< image filename="/images/slow_log.png" alt="Redis Cloud dashboard showing slow database operations" >}}
+
+| Issue | Remediation |
+| ------ | :------ |
+|The KEYS command shows up in the slow log |Find the application that issues the KEYS command and replace it with a SCAN command. In an emergency situation, you can [alter the ACLs for the database user](https://redis.io/docs/latest/operate/rs/security/access-control/rbac/configure-acl/) so that Redis will reject the KEYS command altogether. |
```

Comment:
Couldn't make a relref for this one because I couldn't find it. Don't know if it's moved or obsolete?

Reply:
good catch, fixed